core-common/src/main/java/org/zalando/nakadi/service/KPIEventMapper.java

```diff
@@ -0,0 +1,109 @@
+
+    public KPIEventMapper() {
+    }
+
+    public KPIEventMapper(final Set<Class<? extends KPIEvent>> mappedClasses) {
+        for (final Class<? extends KPIEvent> mappedClass : mappedClasses) {
+            fieldGetterCache.put(mappedClass, findGettersOfKPIFields(mappedClass));
+        }
+    }
+
+    private Set<KPIGetter> findGettersOfKPIFields(final Class<? extends KPIEvent> mappedClass) {
+        return Arrays.stream(mappedClass.getDeclaredFields())
+                .filter(field -> field.isAnnotationPresent(KPIEvent.KPIField.class))
+                .map(field -> findGetterByField(mappedClass, field))
+                .collect(Collectors.toSet());
+    }
+
+    private KPIGetter findGetterByField(final Class<? extends KPIEvent> clazz, final Field field) {
+        final var kpiField = field.getAnnotation(KPIEvent.KPIField.class);
+        final String getterName;
+        if (kpiField.getter().isEmpty()) {
+            final var fieldName = field.getName();
+            getterName = "get" + fieldName.substring(0, 1).toUpperCase() + fieldName.substring(1);
+        } else {
+            getterName = kpiField.getter();
+        }
+        try {
+            final var getter = clazz.getDeclaredMethod(getterName);
+            return new KPIGetter(kpiField.value(), getter);
+        } catch (final NoSuchMethodException nsme) {
+            throw new InternalNakadiException("No getter for " + clazz.getName() + "." + field.getName(), nsme);
+        }
+    }
+
+    public GenericRecord mapToGenericRecord(final KPIEvent kpiEvent, final Schema schema) {
+        final var recordBuilder = new GenericRecordBuilder(schema);
+        gettersOf(kpiEvent.getClass()).forEach(kpiGetter -> {
+            try {
+                recordBuilder.set(kpiGetter.name, kpiGetter.getter.invoke(kpiEvent));
+            } catch (final IllegalAccessException | InvocationTargetException iae) {
+                throw new KPIEventMappingException("Cannot map KPIEvent " + kpiEvent.getClass().getName()
+                        + "to GenericRecord. Unable to read " + kpiGetter.name, iae);
+            }
+        });
+        return recordBuilder.build();
+    }
+
+    public JSONObject mapToJsonObject(final KPIEvent kpiEvent) {
+        final var jsonObject = new JSONObject();
+        gettersOf(kpiEvent.getClass()).forEach(kpiGetter -> {
+            try {
+                jsonObject.put(kpiGetter.name, kpiGetter.getter.invoke(kpiEvent));
+            } catch (final IllegalAccessException | InvocationTargetException iae) {
+                throw new KPIEventMappingException("Cannot map KPIEvent " + kpiEvent.getClass().getName()
+                        + " to JSON. Unable to read " + kpiGetter.name, iae);
+            }
+        });
+        return jsonObject;
+    }
+
+    private Set<KPIGetter> gettersOf(final Class<? extends KPIEvent> clazz) {
+        if (!this.fieldGetterCache.containsKey(clazz)) {
+            synchronized (this) {
```

Comment:
I am not sure why do you have it? we should make sure that we have loaded all possible kpi events before we start, then there will be no need to cache on the fly

Reply:
We plan to load it all during start, but this is there to make the code still work if we forgot to add one.

Reply:
can we avoid that logic for now and fail if the code meets unrecognised bean ?

Reply:
Done